### ISSUE TITLE
`AST` not fully aligned with Babel: `TSClassImplements` used instead of `TSExpressionWithTypeArguments`

### ISSUE THREAD

**Issue by coderaiser** [typescript-eslint/typescript-eslint#4222]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/main/docs/linting/TROUBLESHOOTING.md) and my problem is not listed.

**Repro**

```TS
// the code you're trying to parse
class A implements B {
}

```

```json
{
    "parser": "@typescript-eslint/parser",
    "plugins": [
        "putout",
        "@typescript-eslint"
    ],
    "extends": [
        "plugin:putout/recommended",
        "plugin:@typescript-eslint/recommended"
    ]
}

```

**Expected Result**

`TSExpressionWithTypeArguments` instead of `TSClassImplements` for `implements` node.
 

**Actual Result**

> I work closely with the TypeScript Team and we are gradually aligning the AST of this project with the one produced by Babel's parser. 

As I understood, produced `AST` should be aligned to `Babel`. But it has differences.
`@typescript-eslint/typescript-estree` produces `TSClassImplements` and `Babel` produces `TSExpressionWithTypeArguments` for the same code. This is not a big problem, and it is handled right now with [estree-to-babel](https://github.com/coderaiser/estree-to-babel). 

**Additional Info**

I just want to know is aligning this type of node on roadmap? And what else node types can differ?

Related to:
- https://github.com/coderaiser/putout/issues/90
- https://github.com/EvgenyOrekhov/eslint-config-hardcore/issues/385

By the way, thank you for such a useful tool, you are doing amazing work!

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `latest` |
| `TypeScript`                           | `latest` |
| `node`                                 | `v16` |

**Comment by bradzacher** [comment]
The alignment is in the other direction - we are the canonical typescript-estree representation.

When someone lints their TS code using ESLint - they do so using our tooling. It's very, very rare now-a-days that people use babel to parse their TS code with ESLint. As an example - you can't lint with type information when you use babel.

In the past we have tried to help align their AST with ours to ensure lint rules work - but given that people don't really use use it as much any more the effort has very much dropped off over the years and we only really seek alignment on new features.

**Comment by armano2** [comment]
if you are interested with alignment between babel and typescript-eslint, you should definitely look into those files:
[typescript-estree/tests/ast-alignment/utils.ts](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-estree/tests/ast-alignment/utils.ts) and [typescript-estree/tests/ast-alignment/fixtures-to-test.ts](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-estree/tests/ast-alignment/fixtures-to-test.ts)

first one describes known fixes between babel and typescript and 2nd alignment tests that do not pass for various of reasons

**Comment by coderaiser** [comment]
@armano2 Thanks for quick answer! I’ll give you a little bit more details. The thing is I’m working on linter [🐊`Putout`](https://github.com/coderaiser/putout) that fixes everything it can find. It is built on top of `Babel` because it has the most powerful `API` for `AST`-manipulations and even a [handbook](https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md). 🐊`Putout` has it’s own `CLI` and has first class support of `ESLint`. Also it has a plugin for `ESLint` and can be used in `IDE`, similar to `typescript-eslint`.
But I’d doesn’t provide own parser. It converts `estree` into `Babel` before running as plugin. It made to have ability to parse `vue` and other JavaScript that was get using processors.

So to the point, I agree with you that typings information is very important while listing, I’m OK with `typescript-eslint` parser. One of my users got into trouble so I started research what exactly is a problem.

Thank you for the links, it has useful comments, and of course I can add more conventions into `estree-to-Babel`.

What I want to know is what is the reasoning of using name `TSClassImplements` (https://github.com/babel/babel/issues/12884#issuecomment-785354327). 
Why names so different? What do you think about aligning it to be more compatible, or write the differences like it is made in the [@babel/parser](https://babeljs.io/docs/en/babel-parser#output) page

**Comment by bradzacher** [comment]
> Why names so different?

Because they are two different projects that were written by different groups of people that did not work together in the beginning. And now it's difficult to change the AST because it requires breaking changes in major releases.

And given that it's really, really rare that someone want to be able to use either packages to produce an AST in such a way that they can interop between them - truly standardising is just not something anyone has been motivated enough to do.

---

As mentioned though - we are the canonical "TypeScript ESTree" AST. We have hundreds of ESLint lint rules built on top of our representation - lint rules which directly interrogate the AST.

It's impossible to say exactly how many consumers of the babel TS ESTree representation there are - but I would be confident saying that it's much, much fewer.

**Comment by coderaiser** [comment]
> I work closely with the TypeScript Team and we are gradually aligning the AST of this project with the one produced by Babel's parser.

This is very confusing, I wouldn’t create an issue without this message, so

> And given that it's really, really rare that someone want to be able to use either packages to produce an AST in such a way that they can interop between them

Looks strange… Anyways, thank you for pointing me up to files with very useful comments. It helps a lot, and I’m already ported some nodes to [estree-to-babel](https://github.com/coderaiser/estree-to-babel), would be amazing to have it documented in more readable form :).

Thanks a lot for your work! I also want to mention that it is not a rare case, I’m one of customers, and 🐊`Putout` has simplest possible way of writing [plugins](https://github.com/coderaiser/putout/tree/master/packages/engine-runner#replacer) and  supports types of files other then `JavaScript/TypeScript`, so would be amazing if it will be a little bit simpler to support `AST` produced by `typescript-estree`, because it is useful tool with a lot plugins and great future :).

Also `Babel API` very powerful, so it’s very useful to have support of it.

**Comment by JoshuaKGoldberg** [comment]
This was a super interesting discussion, thanks both of you! 😄 

Marking as a `documentation` issue as we could add this to the FAQs.

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/pull/3258 will be the best form of documentation we can provide. It'll have lists of differences with easy to find snapshots and diffs comparing the ASTs directly.

Just need to spend the time to finish it.

With that PR we'll no longer be manually maintaining exclusion lists which will make it easier to track the differences as babel evolves.

For reference we're are collaborating closely on all new AST features and are ensuring alignment.

> This is very confusing, I wouldn’t create an issue without this message

Yeah things change and priorities shift. There was a bit of a push for alignment but then interest waned as we realised that the world preferred to use our tooling instead of babel to lint.

It's still a goal of ours to have them aligned! But we would need a champion to properly drive the work.

> I also want to mention that it is not a rare case, I’m one of customers,

Sorry I wasn't trying to discount your work. I merely mean that almost all use cases use one or the other - not both.

Your usecase of using both AST reps and converting between them is probably one of the few usecases that exists

As an example - all lint rules are built against our ast spec and do not consider babel's nodes at all - I don't know of any plugins which are built to handle the differences between the two.

**Comment by JoshuaKGoldberg** [comment]
Closing out as #3258 was resolved and in the >2.5 years since this was filed, nobody else has +1d. If anybody has any specific docs requests please do file a new issue explaining what they are and why. Cheers all!
